src/components/layouts/llama-thumbnail-grid-layout.ts

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2022 Google Inc. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import { customElement } from 'lit/decorators.js';
+
+import { LlamaVerticalScrollLayout } from './llama-vertical-scroll-layout';
+
+@customElement('llama-thumbnail-grid-layout')
+/**
+ * Renders all items at once (downscaled) in a dynamically-sized grid.
+ */
+export class LlamaThumbnailGridLayout extends LlamaVerticalScrollLayout {
+  // TODO: it would be nice to specify the CSS as part of the element, rather than in the global
+  // css file - otherwise the only point of this class is to get the element name in the DOM.
+  // I'm sure there's got to be a way to make the llama-item children work even if this element
+  // has a shadow root (@import url(../llaminator.css) _almost_ works).
```

Comment:
Would having `llama-thumbnail-grid-layout.css` work w/ an `import` statement in this file? I _think_ that our build system is setup to automatically concatenate such imports into the main CSS file.

Reply:
well time for another rant. I spent the past not-quite-sure-how-long frustratingly trying to import npmjs.com/package/material-icons directly into a particular shadow-dom custom element, and gotten nowhere.

[From what I can tell](https://lit.dev/docs/components/styles/#styles-in-the-template), if you don't use global styles, the "right way" to specify css is directly in [lit `css` tags](https://lit.dev/docs/components/styles/#add-styles), as adding them in the `html` templates will cause them to get loaded per-element and be inefficient and FOUC.

However, no [lit documentation](https://lit.dev/docs/components/styles/) I can find mentions anything about importing css into a `css` tag template. I believe (at some point) I tried importing with regular css `@import url` commands, but couldn't get that to work.

`import 'material-icons/iconfont/filled.css';` (assuming webpack is set up to use [css-loader](https://webpack.js.org/loaders/css-loader/) (_and_ `style-loader` - it's still not clear to me why there need to be 2), then it'll take care of importing things like the button icon "fonts" into the project, and making the proper styles globally available on the page (I think), but it doesn't provide a way to use it only within the scope of a custom element - the `import` statement doesn't give you access to the actual contents of the CSS file from within js/ts (doesn't seem like that should be so difficult).

I also tried looking into using [css modules and/or the MiniCssExtractPlugin plugin](https://blog.jakoblind.no/css-modules-webpack/), but neither of those really addressed the problem (I would still need to import the styles using a `<link>` or `<style>` in the `html` template).

I also considered trying out [raw-loader](https://v4.webpack.js.org/loaders/raw-loader/), but IIUC, using that instead of `css-loader`+`style-loader` means that none of the imports inside the CSS file will get resolved properly?

Looking a bit deeper, it's possible that changing [css-loader's exportType](https://webpack.js.org/loaders/css-loader/#options) might help? Kind of at the point of giving up for now, though.

Just for the sake of completeness, [material-components-web's documentation](https://github.com/material-components/material-components-web/tree/master/packages/mdc-button) also didn't really help at all.

At least [I'm not the only one](https://forum.freecodecamp.org/t/using-external-css-in-a-web-component-with-webpack/326307) who's gone through this, but I don't really understand their solution.

```
import styles from './main.style.css';
```

what is in `main.style.css`? I tried that on my css file and it complained because there was no `styles` to import from not-a-javascript-file. Seems they're using the same loaders I tried... Is `main.style.css` already a javascript file that contains css declarations? If I knew how to generate that file as part of the build, then I would be able to import it easily enough.

Reply:
I'll look into making the css loader work with open DOM in a follow-up PR, I give up on shadow dom